### ISSUE TITLE
[Discussion] bool CanX hook short-circuiting.

### ISSUE THREAD

**Issue by Mirsario** [tModLoader/tModLoader#1079]
I have noticed that, currently, hook functions like ItemLoader.CanUseItem do not use short-circuiting on purpose, and have not found any reason as to why.

In my and some others' opinion, all **non-nullable bool** CanX hooks should short-circuit for the sake of performance, while hooks like PreX (PreAI) should not, as their timing is useful for custom logic, not always related to the return value.

I am looking to hear if anyone had found some non-orthodox uses for CanUseItem or something similar, to see if there should be some new hook made for that, after short-circuiting is enabled for such methods.

**Comment by hamstar0** [comment]
Yes please. Also, PreAI and CanAI should both still exist, though PreAI should have no return value.

**Comment by Mirsario** [comment]
Semi-implemented, but we should check every hook.
